aws/rust-runtime/aws-types/src/environment.rs

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Abstractions for testing code that interacts with the process environment:
+//! - Determining the current time
+//! - Reading environment variables
+//! - Reading from the file system
+
+use std::collections::HashMap;
+use std::env::VarError;
+use std::ffi::OsString;
+use std::path::Path;
+use std::sync::Arc;
+
+/// FileSystem abstraction
+///
+/// Simple filesystem abstraction enabling in-memory mocking of the filesystem
+///
+/// # Example
+/// Construct a file-system which delegates to `std::fs`:
+/// ```rust
+/// let fs = aws_types::environment::Fs::real();
+/// ```
+///
+/// Construct an in-memory file system for testing:
+/// ```rust
+/// use std::collections::HashMap;let fs = aws_types::environment::Fs::from_map({
+///     let mut map = HashMap::new();
+///     map.insert("/home/.aws/config".to_string(),"[default]\nregion = us-east-1".into());
+///     map
+/// });
+/// ```
+pub struct Fs(fs::Inner);
+
+impl Fs {
+    pub fn real() -> Self {
+        Fs(fs::Inner::Real)
+    }
+
+    pub fn from_raw_map(fs: HashMap<OsString, Vec<u8>>) -> Self {
+        Fs(fs::Inner::Fake { fs })
+    }
+
+    pub fn from_map(data: HashMap<String, Vec<u8>>) -> Self {
+        let fs = data.into_iter().map(|(k, v)| (k.into(), v)).collect();
+        Fs(fs::Inner::Fake { fs })
+    }
+
+    pub fn exists(&self, path: impl AsRef<Path>) -> bool {
+        use fs::Inner;
+        let path = path.as_ref();
+        match &self.0 {
+            Inner::Real => path.exists(),
+            Inner::Fake { fs, .. } => fs.contains_key(path.as_os_str()),
+        }
+    }
+
+    pub fn read(&self, path: impl AsRef<Path>) -> std::io::Result<Vec<u8>> {
+        use fs::Inner;
+        let path = path.as_ref();
+        match &self.0 {
+            Inner::Real => std::fs::read(path),
+            Inner::Fake { fs } => fs
+                .get(path.as_os_str())
+                .cloned()
+                .ok_or_else(|| std::io::ErrorKind::NotFound.into()),
+        }
+    }
+}
+
+mod fs {
+    use std::collections::HashMap;
+    use std::ffi::OsString;
+
+    pub enum Inner {
+        Real,
+        Fake { fs: HashMap<OsString, Vec<u8>> },
+    }
+}
+
+/// Process Environment Abstraction
```

Comment:
Recommend adding more details on what this means. I had to read the code to realize it's an abstraction over environment variables.